mypyc 1.16 regression: x86_64-cpython-3{9,10,11,12,13} error: ‘FOO_finalize’ undeclared here (not in a function);  did you mean ‘FOO_init’?

**Crash Report**
compiling Python modules using mypyc failed with a gcc error, where it did not using mypy 1.15

Python 3.9: https://github.com/common-workflow-language/schema_salad/actions/runs/15347015751/job/43185437280?pr=927#step:6:450

Python 3.10: https://github.com/common-workflow-language/schema_salad/actions/runs/15347015751/job/43185437282?pr=927#step:6:439

Python 3.11: https://github.com/common-workflow-language/schema_salad/actions/runs/15347015751/job/43185437303?pr=927#step:6:433

Python 3.12: https://github.com/common-workflow-language/schema_salad/actions/runs/15347015751/job/43185437295?pr=927#step:6:425

Python 3.13: https://github.com/common-workflow-language/schema_salad/actions/runs/15347015751/job/43185437315?pr=927#step:6:425

**Traceback**

```
    building '5df31530eea7292b5dae__mypyc' extension
    creating build/temp.linux-x86_64-cpython-39/build
    gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -I/tmp/pip-build-env-qntpj1xx/overlay/lib/python3.9/site-packages/mypyc/lib-rt -Ibuild -I/opt/_internal/cpython-3.9.22/include/python3.9 -c build/__native_5df31530eea7292b5dae.c -o build/temp.linux-x86_64-cpython-39/build/__native_5df31530eea7292b5dae.o -O3 -g0 -Werror -Wno-unused-function -Wno-unused-label -Wno-unreachable-code -Wno-unused-variable -Wno-unused-command-line-argument -Wno-unknown-warning-option -Wno-unused-but-set-variable -Wno-ignored-optimization-argument -Wno-cpp
    build/__native_5df31530eea7292b5dae.c:110530:32: error: ‘ref_resolver___NormDict_finalize’ undeclared here (not in a function); did you mean ‘ref_resolver___NormDict_init’?
    110530 |     .tp_finalize = (destructor)ref_resolver___NormDict_finalize,
           |                                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           |                                ref_resolver___NormDict_init
    cc1: note: unrecognized command-line option ‘-Wno-ignored-optimization-argument’ may have been intended to silence earlier diagnostics
    cc1: note: unrecognized command-line option ‘-Wno-unknown-warning-option’ may have been intended to silence earlier diagnostics
    cc1: note: unrecognized command-line option ‘-Wno-unused-command-line-argument’ may have been intended to silence earlier diagnostics
    error: command '/opt/rh/gcc-toolset-14/root/usr/bin/gcc' failed with exit code 1
    error: subprocess-exited-with-error
    
    × Building wheel for schema-salad (pyproject.toml) did not run successfully.
    │ exit code: 1
    ╰─> See above for output.
```

**To Reproduce**

1. Checkout https://github.com/common-workflow-language/schema_salad/tree/dependabot/pip/mypy-mypyc--1.16.0
2. `tox -e py39-mypy` (or `py310-mypy`, up to `py313-mypy`); or make a virtualenv and activate it, and `make dev mypyc`

**Your Environment**

<!-- Include as many relevant details about the environment you experienced the bug in -->

- Mypy version used: 1.16
- Mypy command-line flags: `MYPYPATH=mypy-stubs SCHEMA_SALAD_USE_MYPYC=1 pip install --verbose -e .` 
- Mypy configuration options from `mypy.ini` (and other config files): https://github.com/common-workflow-language/schema_salad/blob/dependabot/pip/mypy-mypyc--1.16.0/mypy.ini
- Python version used: 3.9 - 3.13
- Operating system and version: linux

I nerdsniped myself, here is a minimal reproducer

`ref_resolver.py`
``` python
from typing import Any, Callable, Union

from ruamel.yaml.comments import CommentedMap, CommentedSeq

class NormDict(dict[str, Union[CommentedMap, CommentedSeq, str, None]]):
    """A Dict where all keys are normalized using the provided function."""

    def __init__(self, normalize: Callable[[str], str] = str) -> None:
        super().__init__()
        self.normalize = normalize

    def __eq__(self, other: Any) -> bool:
        return super().__eq__(other)

    def __getitem__(self, key: Any) -> Any:
        return super().__getitem__(self.normalize(key))

    def __setitem__(self, key: Any, value: Any) -> Any:
        return super().__setitem__(self.normalize(key), value)

    def __delitem__(self, key: Any) -> Any:
        return super().__delitem__(self.normalize(key))

    def __contains__(self, key: Any) -> bool:
        return super().__contains__(self.normalize(key))

    def __del__(self) -> None:
        del self.normalize
```
``` bash
$ pip install mypy==1.15.0
[…]
$ mypyc ref_resolver.py 
running build_ext
building 'ref_resolver' extension
creating build/temp.linux-x86_64-cpython-311
creating build/temp.linux-x86_64-cpython-311/build
x86_64-linux-gnu-gcc -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -fstack-clash-protection -Wformat -Werror=format-security -fcf-protection -g -fwrapv -O2 -fPIC -I/home/michael/ss2/env3.11/lib/python3.11/site-packages/mypyc/lib-rt -I/home/michael/ss2/env3.11/include -I/usr/include/python3.11 -c build/__native.c -o build/temp.linux-x86_64-cpython-311/build/__native.o -O3 -g1 -Werror -Wno-unused-function -Wno-unused-label -Wno-unreachable-code -Wno-unused-variable -Wno-unused-command-line-argument -Wno-unknown-warning-option -Wno-unused-but-set-variable -Wno-ignored-optimization-argument -Wno-cpp
creating build/lib.linux-x86_64-cpython-311
x86_64-linux-gnu-gcc -shared -Wl,-O1 -Wl,-Bsymbolic-functions -g -fwrapv -O2 build/temp.linux-x86_64-cpython-311/build/__native.o -L/usr/lib/x86_64-linux-gnu -o build/lib.linux-x86_64-cpython-311/ref_resolver.cpython-311-x86_64-linux-gnu.so
copying build/lib.linux-x86_64-cpython-311/ref_resolver.cpython-311-x86_64-linux-gnu.so ->

$ rm -Rf ref_resolver.*.so .mypy_cache/ build/
$ pip install mypy==1.16.0
[…]
$ mypyc ref_resolver.py 
running build_ext
building 'ref_resolver' extension
creating build/temp.linux-x86_64-cpython-311
creating build/temp.linux-x86_64-cpython-311/build
x86_64-linux-gnu-gcc -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -fstack-clash-protection -Wformat -Werror=format-security -fcf-protection -g -fwrapv -O2 -fPIC -I/home/michael/ss2/env3.11/lib/python3.11/site-packages/mypyc/lib-rt -I/home/michael/ss2/env3.11/include -I/usr/include/python3.11 -c build/__native.c -o build/temp.linux-x86_64-cpython-311/build/__native.o -O3 -g1 -Werror -Wno-unused-function -Wno-unused-label -Wno-unreachable-code -Wno-unused-variable -Wno-unused-command-line-argument -Wno-unknown-warning-option -Wno-unused-but-set-variable -Wno-ignored-optimization-argument -Wno-cpp
build/__native.c:162:32: error: ‘NormDict_finalize’ undeclared here (not in a function); did you mean ‘NormDict_init’?
  162 |     .tp_finalize = (destructor)NormDict_finalize,
      |                                ^~~~~~~~~~~~~~~~~
      |                                NormDict_init
cc1: note: unrecognized command-line option ‘-Wno-ignored-optimization-argument’ may have been intended to silence earlier diagnostics
cc1: note: unrecognized command-line option ‘-Wno-unknown-warning-option’ may have been intended to silence earlier diagnostics
cc1: note: unrecognized command-line option ‘-Wno-unused-command-line-argument’ may have been intended to silence earlier diagnostics
error: command '/usr/lib/ccache/x86_64-linux-gnu-gcc' failed with exit code 1
$ x86_64-linux-gnu-gcc --version
x86_64-linux-gnu-gcc (Debian 12.2.0-14+deb12u1) 12.2.0
```

--- a/mypyc/codegen/emitclass.py
+++ b/mypyc/codegen/emitclass.py
@@ -214,12 +214,14 @@ def generate_class(cl: ClassIR, module: str, emitter: Emitter) -> None:
     if not cl.builtin_base:
         fields["tp_new"] = new_name
 
-    if generate_full:
-        fields["tp_dealloc"] = f"(destructor){name_prefix}_dealloc"
-        fields["tp_traverse"] = f"(traverseproc){name_prefix}_traverse"
-        fields["tp_clear"] = f"(inquiry){name_prefix}_clear"
     # Populate .tp_finalize and generate a finalize method only if __del__ is defined for this class.
     del_method = next((e.method for e in cl.vtable_entries if e.name == "__del__"), None)
+    # Del method is called from dealloc method. So, dealloc should be generated whenever there is a del method.
+    if generate_full or del_method:
+        fields["tp_dealloc"] = f"(destructor){name_prefix}_dealloc"
+    if generate_full:
+        fields["tp_traverse"] = f"(traverseproc){name_prefix}_traverse"
+        fields["tp_clear"] = f"(inquiry){name_prefix}_clear"
     if del_method:
         fields["tp_finalize"] = f"(destructor){finalize_name}"
     if needs_getseters:
@@ -300,13 +302,6 @@ def generate_class(cl: ClassIR, module: str, emitter: Emitter) -> None:
         emit_line()
         generate_traverse_for_class(cl, traverse_name, emitter)
         emit_line()
-        generate_clear_for_class(cl, clear_name, emitter)
-        emit_line()
-        generate_dealloc_for_class(cl, dealloc_name, clear_name, bool(del_method), emitter)
-        emit_line()
-        if del_method:
-            generate_finalize_for_class(del_method, finalize_name, emitter)
-            emit_line()
 
         if cl.allow_interpreted_subclasses:
             shadow_vtable_name: str | None = generate_vtables(
@@ -316,6 +311,14 @@ def generate_class(cl: ClassIR, module: str, emitter: Emitter) -> None:
         else:
             shadow_vtable_name = None
         vtable_name = generate_vtables(cl, vtable_setup_name, vtable_name, emitter, shadow=False)
+        emit_line()
+    if generate_full or del_method:
+        generate_clear_for_class(cl, clear_name, emitter)
+        emit_line()
+        generate_dealloc_for_class(cl, dealloc_name, clear_name, bool(del_method), emitter)
+        emit_line()
+    if del_method:
+        generate_finalize_for_class(del_method, finalize_name, emitter)
         emit_line()
     if needs_getseters:
         generate_getseter_declarations(cl, emitter)

--- a/mypyc/codegen/emitclass.py
+++ b/mypyc/codegen/emitclass.py
@@ -214,14 +214,12 @@ def generate_class(cl: ClassIR, module: str, emitter: Emitter) -> None:
     if not cl.builtin_base:
         fields["tp_new"] = new_name
 
-    # Populate .tp_finalize and generate a finalize method only if __del__ is defined for this class.
-    del_method = next((e.method for e in cl.vtable_entries if e.name == "__del__"), None)
-    # Del method is called from dealloc method. So, dealloc should be generated whenever there is a del method.
-    if generate_full or del_method:
-        fields["tp_dealloc"] = f"(destructor){name_prefix}_dealloc"
     if generate_full:
+        fields["tp_dealloc"] = f"(destructor){name_prefix}_dealloc"
         fields["tp_traverse"] = f"(traverseproc){name_prefix}_traverse"
         fields["tp_clear"] = f"(inquiry){name_prefix}_clear"
+    # Populate .tp_finalize and generate a finalize method only if __del__ is defined for this class.
+    del_method = next((e.method for e in cl.vtable_entries if e.name == "__del__"), None)
     if del_method:
         fields["tp_finalize"] = f"(destructor){finalize_name}"
     if needs_getseters:
@@ -302,6 +300,10 @@ def generate_class(cl: ClassIR, module: str, emitter: Emitter) -> None:
         emit_line()
         generate_traverse_for_class(cl, traverse_name, emitter)
         emit_line()
+        generate_clear_for_class(cl, clear_name, emitter)
+        emit_line()
+        generate_dealloc_for_class(cl, dealloc_name, clear_name, bool(del_method), emitter)
+        emit_line()
 
         if cl.allow_interpreted_subclasses:
             shadow_vtable_name: str | None = generate_vtables(
@@ -311,11 +313,6 @@ def generate_class(cl: ClassIR, module: str, emitter: Emitter) -> None:
         else:
             shadow_vtable_name = None
         vtable_name = generate_vtables(cl, vtable_setup_name, vtable_name, emitter, shadow=False)
-        emit_line()
-    if generate_full or del_method:
-        generate_clear_for_class(cl, clear_name, emitter)
-        emit_line()
-        generate_dealloc_for_class(cl, dealloc_name, clear_name, bool(del_method), emitter)
         emit_line()
     if del_method:
         generate_finalize_for_class(del_method, finalize_name, emitter)

--- a/mypyc/test-data/run-classes.test
+++ b/mypyc/test-data/run-classes.test
@@ -2754,7 +2754,10 @@ def test_function():
     assert(isinstance(d.fitem, ForwardDefinedClass))
     assert(isinstance(d.fitems, ForwardDefinedClass))
 
-[case testDelForDictSubclass]
+[case testDelForDictSubclass-xfail]
+# The crash in issue mypy#19175 is fixed.
+# But, for classes that derive from built-in Python classes, user-defined __del__ method is not
+# being invoked.
 class DictSubclass(dict):
     def __del__(self):
         print("deleting DictSubclass...")
@@ -2785,6 +2788,12 @@ class C(B):
 
 class D(A):
     pass
+
+# Just make sure that this class compiles (see issue mypy#19175). testDelForDictSubclass tests for
+# correct output.
+class NormDict(dict):
+    def __del__(self) -> None:
+        pass
 
 [file driver.py]
 import native

--- a/mypyc/test-data/run-classes.test
+++ b/mypyc/test-data/run-classes.test
@@ -2754,6 +2754,18 @@ def test_function():
     assert(isinstance(d.fitem, ForwardDefinedClass))
     assert(isinstance(d.fitems, ForwardDefinedClass))
 
+[case testDelForDictSubclass]
+class DictSubclass(dict):
+    def __del__(self):
+        print("deleting DictSubclass...")
+
+[file driver.py]
+import native
+native.DictSubclass()
+
+[out]
+deleting DictSubclass...
+
 [case testDel]
 class A:
     def __del__(self):

